src/applications/mesh_deformation/cube.c

```diff
@@ -0,0 +1,153 @@
+ * other values correspond to static values of dimensionality.
+ * will be replace by proposed values. If you want to use this file outside
+ * STARS-H, simply do substitutions yourself.
+ *
+ * @file src/applications/mesh_deformation/cube.c
+ * @version 0.1.1
+ * @author Rabab Alomairy
+ * @date 2020-05-09
+ */
+
+#include "common.h"
+#include "starsh.h"
+#include "starsh-rbf.h"
+#include <inttypes.h>
+#include <math.h>
+#include <stdio.h>
+
+
+
+/*! Computing cube Coordinates
+ * @param[inout] v: Mesh Coordinates
+ * @param[in] index: element position
+ * @param[in] L: edge length of cube
+ * @param[in] n:  overall size of cube //TODO ask application pp
+*/
+
+static void cube(double* v, int index, double L, int n)
+{
+        double step = 1 / (double)(n - 1);
+        double x = -1;
+        double y = -1;
+        double z = -1;
+
+        if (index < 2 * pow(n, 2))
+        {
+                z = index / (int)pow(n, 2);
+                int ind = index - z*pow(n, 2);
+
+                x = (ind / n)* step;
+                y = (ind % n)* step;
+        }
+        else if ((index >= 2 * pow(n, 2)) && (index < 4 * (pow(n, 2) - n)))
+        {
+                int ind = index - 2 * pow(n, 2);
+                x = ind / (int)(pow(n, 2) - 2 * n);
+                ind =(int)(ind - x*(pow(n, 2) - 2 * n));
+
+                y = (ind % n)* step;
+                z = ((ind / n) + 1)* step;
+        }
+        else if ((index >= 4 * (pow(n, 2) - n)) && (index < 6 * pow(n, 2) - 12 * n + 8))
+        {
+                int ind = index - 4 * (pow(n, 2) - n);
+                y = ind / (int)(pow(n, 2) - 4 * n + 4);
+                ind =(int)(ind - y*(pow(n, 2) - 4 * n + 4));
+
+                x = ((ind / (n - 2)) + 1)* step;
+                z = ((ind % (n - 2)) + 1)* step;
+        }
+        v[0] = x*L;
+        v[1] = y*L;
+        v[2] = z*L;
+
+}
+
+/*! Fills matrix \f$ A \f$ with values
+ * \f[
+ *      A_{ij} = \frac{1}{r_{ij}},
+ * \f]
+ * \f$ r_{ij} \f$ is a distance between \f$i\f$-th and \f$j\f$-th mesh
+ * points. No memory is allocated in this function!
+ *
+ * @param[in] nrows: Number of rows of \f$ A \f$.
+ * @param[in] ncols: Number of columns of \f$ A \f$.
+ * @param[in] irow: Array of row indexes.
+ * @param[in] icol: Array of column indexes.
+ * @param[in] row_data: Pointer to physical data (\ref STARSH_mddata object).
+ * @param[in] col_data: Pointer to physical data (\ref STARSH_mddata object).
+ * @param[out] result: Pointer to memory of \f$ A \f$.
+ * @param[in] ld: Leading dimension of `result`.
+ * */
+
+void starsh_generate_3d_cube(int nrows, int ncols,
+        STARSH_int *irow, STARSH_int *icol, void *row_data, void *col_data,
+        void *result, int lda)
+{
+        double vi[3];
+        double vj[3];
+        STARSH_mddata *data = row_data;
+        STARSH_int i0, j0;
+        int n = floor(1 + pow(1 - (8 - data->mesh_points) / (double)6, 0.5)) + 1;
```

Comment:
I would suggest using sqrt function for pow(x, 0.5)

Reply:
In addition to that, look at https://stackoverflow.com/questions/18155883/strange-behaviour-of-the-pow-function
`pow` function works with floating numbers and can round result down, e.g. `pow(100, 2)=9999.9999` and your `floor` function will then return 9999 instead of 10000. Remedy is to add 0.5 before doing `floor` operation.